modules/affile/affile-dest.c

```diff
@@ -162,25 +163,105 @@ affile_dw_reopen(AFFileDestWriter *self)
 
       LogTransport *transport = file_opener_construct_transport(self->owner->file_opener, fd);
 
-      proto = file_opener_construct_dst_proto(self->owner->file_opener, transport,
-                                              &self->owner->writer_options.proto_options);
-    }
-  else if (open_result == FILE_OPENER_RESULT_ERROR_PERMANENT)
-    {
-      return FALSE;
+      *p = file_opener_construct_dst_proto(self->owner->file_opener, transport,
+                                           &self->owner->writer_options.proto_options);
+
+      // get current filesize
+      if (fstat(fd, &st) == 0)
+        {
+          self->cached_filesize = st.st_size;
+        }
+      else
+        {
+          msg_error("Error reading file size after opening file",
+                    evt_tag_str("filename", self->filename),
+                    evt_tag_errno("errno", errno));
+          // treating as opening error
+          open_result = FILE_OPENER_RESULT_ERROR_TRANSIENT;
+        }
     }
-  else
+  else if (open_result == FILE_OPENER_RESULT_ERROR_TRANSIENT)
     {
       msg_error("Error opening file for writing",
                 evt_tag_str("filename", self->filename),
                 evt_tag_error(EVT_TAG_OSERROR));
     }
 
+  return open_result;
+}
+
+static gboolean
+affile_dw_reopen(AFFileDestWriter *self)
+{
+  LogProtoClient *proto = NULL;
+  FileOpenerResult open_result = _dw_reopen(self, &proto);
+
+  if (open_result == FILE_OPENER_RESULT_ERROR_PERMANENT)
+    {
+      return FALSE;
+    }
+
   log_writer_reopen(self->writer, proto);
 
   return TRUE;
 }
 
+/*
+ * This function checks if logrotation should be performed based on the current
+ * options and the filename. After rotating the logs the current logfile is
+ * reopened and the new fd returned through the pointer arg.
+ *
+ * Alternativly the reopened could be triggered through a main-thread call here:
+ *
+ *      main_loop_call((MainLoopTaskFunc) affile_dw_reopen, (gpointer) self, TRUE);
+ *
+ */
+static gboolean
+affile_dw_logrotate(AFFileDestWriter *self, gpointer user_data)
+{
+  gpointer *args = (gpointer *) user_data;
+  LogProtoClient **p = (LogProtoClient **) args[0];
+  const gsize buf_len = (gsize) args[1];
+
+  LogRotateOptions *logrotate_options = &(self->owner->logrotate_options);
+  const gchar *filename = self->filename;
+
+  self->cached_filesize += buf_len;
+
+  if (is_logrotate_enabled(logrotate_options) && is_logrotate_pending(logrotate_options, self->cached_filesize))
```

Comment:
actually this must be 3 conditions

enabled, required, pending

the current pending must be called required

a future, new pending must really check if there was a pending, whatever reason not finished rotation in progress

an already in progress rotation, especially the one that already rotated the file names but still waiting for an earlier failed re_open, must not rotate again

please see bellow how the latter can still happen currently